File: scala/bob/src/main/scala/Bob.scala
class Bob {
  val Responses = Map(
    'shouting -> "Whoa, chill out!",
    'silence -> "Fine. Be that way!",
    'question -> "Sure.",
    'other -> "Whatever."
  )

  def hey(speech: String): String = Responses(SpeechParser.typeOfSpeech(speech))
}

object SpeechParser {
  def typeOfSpeech(speech: String): Symbol = {
    if (isShouting(speech))
      'shouting
    else if (isSilence(speech))
      'silence
    else if (isQuestion(speech))
      'question
    else
      'other
  }

  private def isShouting(phrase: String): Boolean = {
    phrase == phrase.toUpperCase && phrase.exists(_.isLetter)
  }

  private def isSilence(phrase: String): Boolean = {
    phrase.trim.isEmpty
  }

  private def isQuestion(phrase: String): Boolean = {
    phrase.endsWith("?")
  }
}

File: python/bob/bob.py
import re

def hey(phrase):
  phrase = phrase.strip()
  if is_silence(phrase):
    return 'Fine. Be that way!'
  elif is_shouting(phrase):
    return "Whoa, chill out!"
  elif is_question(phrase):
    return "Sure."
  return "Whatever."

def is_silence(phrase):
  return phrase == ''

def is_shouting(phrase):
  return re.search('[a-zA-Z]', phrase) and phrase == phrase.upper()

def is_question(phrase):
  return phrase.endswith('?')

